src/Raven.Server/Documents/TransactionCommands/JsonPatchCommand.cs

```diff
@@ -0,0 +1,373 @@
+        }
+        
+        public record Command(CommandTypes Type, string[] Paths, string[] FromPaths, string OriginalPath, string OriginalFrom, object Value)
+        {
+            public BlittableJsonReaderBase GetActionableObject(Document document, string[] paths, string fullPath)
+            {
+                BlittableJsonReaderBase root = document.Data;
+                
+                for (int i = 1; i < paths.Length - 1; i++)
+                {
+                    TryGetMember(root, paths[i], out var member, fullPath);
+                    
+                    if (member is BlittableJsonReaderObject nested)
+                        root = nested;
+                    else if (member is BlittableJsonReaderArray arr)
+                    {
+                        if(i == paths.Length - 2)
+                            return arr;
+                        root = arr;
+                    }
+                    else
+                        throw new ArgumentException($"Cannot reach target location. Failed to fetch '{paths[i+1]}' in path '{fullPath}' for operation '{Type}' because element at {paths[i]} is not an object or a collection");
+                }
+
+                return root;
+            }
+
+            public object GetNonActionableObject(Document document, string[] paths, string fullPath)
+            {
+                var cur = GetActionableObject(document, paths, fullPath);
+                TryGetMember(cur, paths[^1], out object obj, fullPath);
+                return obj;
+            }
+
+            private void TryGetMember(BlittableJsonReaderBase objectToSearchIn, string name, out object result, string fullPath)
+            {
+                switch (objectToSearchIn)
+                {
+                    case BlittableJsonReaderArray arr:
+                        if (int.TryParse(name, out int index) == false)
+                            throw new ArgumentException($"Expected an index but got a '{name}' at path {fullPath}");
+                        if (index >= arr.Length || index < 0)
+                            throw new ArgumentOutOfRangeException($"Position {index} is out of array bounds at array {fullPath}");
+                        result = arr[index];
+                        break;
+
+                    case BlittableJsonReaderObject blittable:
+                        if (blittable.TryGetMember(name, out object obj) == false)
+                        {
+                            goto default;
+                        }
+                        result = obj;
+                        break;
+                    default:
+                        throw new ArgumentException($"Cannot reach target location. Failed to fetch '{name}' in path '{fullPath}' for operation '{Type}'");
+                }
+            }
+
+            public void Test(object objectAtPath, object value)
+            {
+                //No need to check type of objectAtPath because all JsonReader objects implement Equals
```

Comment:
Are you operating on `JsonReader` here, pretty sure that this isn't the case.
What if you are trying to compare `BlitableJsonObject` ?